src/datumaro/plugins/sam_transforms/automatic_mask_gen.py

```diff
@@ -0,0 +1,181 @@
+            launcher_cls = TritonLauncher
+        else:
+            raise ValueError(inference_server_type)
+
+        self._sam_encoder_launcher = launcher_cls(
+            model_name="sam_encoder",
+            model_interpreter_path=osp.abspath(sam_encoder_interp.__file__),
+            model_version=1,
+            host=host,
+            port=port,
+            timeout=timeout,
+            tls_config=tls_config,
+            protocol_type=protocol_type,
+        )
+        self._sam_decoder_launcher = launcher_cls(
+            model_name="sam_decoder",
+            model_interpreter_path=osp.abspath(sam_decoder_for_amg.__file__),
+            model_version=1,
+            host=host,
+            port=port,
+            timeout=timeout,
+            tls_config=tls_config,
+            protocol_type=protocol_type,
+        )
+
+        self.points_per_side = points_per_side
+        self.points_per_batch = points_per_batch
+        self.mask_threshold = mask_threshold
+        self.pred_iou_thresh = pred_iou_thresh
+        self.stability_score_offset = stability_score_offset
+        self.stability_score_thresh = stability_score_thresh
+        self.box_nms_thresh = box_nms_thresh
+        self.min_mask_region_area = min_mask_region_area
+
+        super().__init__(
+            extractor,
+            launcher=self._sam_encoder_launcher,
+            batch_size=1,
+            append_annotation=False,
+            num_workers=num_workers,
+        )
+
+    @property
+    def points_per_side(self) -> int:
+        return self._points_per_side
+
+    @points_per_side.setter
+    def points_per_side(self, points_per_side: int) -> None:
+        points_y = (np.arange(points_per_side) + 0.5) / points_per_side
+        points_x = (np.arange(points_per_side) + 0.5) / points_per_side
+
+        points_x = np.tile(points_x[None, :], (points_per_side, 1))
+        points_y = np.tile(points_y[:, None], (1, points_per_side))
+        self._points_grid = np.stack([points_x, points_y], axis=-1).reshape(-1, 2)
+        self._points_per_side = points_per_side
+
+    def _process_batch(
+        self,
+        batch: List[DatasetItem],
+    ) -> List[DatasetItem]:
+        img_embeds = self._sam_encoder_launcher.launch(
+            batch=[item for item in batch if self._sam_encoder_launcher.type_check(item)]
+        )
+
+        items = []
+        for item, img_embed in zip(batch, img_embeds):
+            amg_masks: List[AMGMasks] = []
+
+            for i in range(0, len(self._points_grid), self.points_per_batch):
+                amg_points = [AMGPoints(points=self._points_grid[i : i + self.points_per_batch])]
+                item_to_decode = item.wrap(annotations=amg_points + img_embed)
+
+                # Nested list of mask [[mask_0, ...]]
+                nested_masks: List[List[AMGMasks]] = self._sam_decoder_launcher.launch(
+                    [item_to_decode],
+                    stack=False,
+                )
```

Comment:
Can I ask why nested list is required? Is this related to `[item_to_decode]`? Finally, we only need to have only `nested_masks[0]`.

Reply:
Although we restrict it to give a single item, `Launcher.launch(x)` assumes `x` is a batch so that it requires `x` to be `list`. Thus, the output should be a nested list.